modules/order/src/AvailabilityOrderProcessor.php

```diff
@@ -37,10 +38,23 @@ public function process(OrderInterface $order) {
     foreach ($order->getItems() as $order_item) {
       $purchased_entity = $order_item->getPurchasedEntity();
       if ($purchased_entity) {
-        $available = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
-        if (!$available) {
-          $order->removeItem($order_item);
-          $order_item->delete();
+        /** @var \Drupal\commerce\AvailabilityResponseInterface $availability */
+        $availability = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
+        if (!$availability->isNeutral()) {
+          if ($availability->getMax() == 0) {
+            $order->removeItem($order_item);
+            $order_item->delete();
+            drupal_set_message(t('The item %item is no longer available and has been removed from your cart.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+          }
+          elseif ($availability->getMax() < $order_item->getQuantity()) {
+            drupal_set_message(new PluralTranslatableMarkup(abs($availability->getMax() - $order_item->getQuantity()), '1 %item is currently unavailable. The quantity in your cart has been updated.', '@count %item are currently unavailable. The quantity in your cart has been updated.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+            $order_item->setQuantity($availability->getMax());
+            $order_item->save();
+          }
```

Comment:
Ah, you're right - this can work! I reviewed the sandbox module and it does look like it'll need some changes in commerce_stock such as the StockCheckInterface::getOrderStockTransactionSum addition, but I think we're headed in a good direction.  I'd open a PR against commerce_stock for backorderable as a submodule, that way you can start writing tests against the commerce_stock for the changes to stock and the addition/integration of backorderable.  I think this PR is ready for a real review from @bojanz.